### PR TITLE
test(jest): Replace jest snapshots with image snapshots

### PR DESCRIPTION
**Pull request by billyvg**
This replaces jest snapshots with rendered image snapshots. You can test this locally with either of these env vars: `VISUAL_SNAPSHOT_ENABLE` or `VISUAL_HTML_ENABLE` - the former will be slower since it needs puppeteer to render the images, the later will output HTML to disk.

This has the benefits of 1) providing useful snapshots (e.g. what visual impact a css change makes rather than seeing a class name change) 2) will eliminate build errors due to outdated snapshots in getsentry (although this raises the question of how we can ensure that we do not introduce regressions in getsentry) 

You can view the rendered snapshots here: https://storage.googleapis.com/sentry-visual-snapshots/getsentry/sentry/c3cb60e108b5c9d83e0dd62c936992dd116982db/index.html#added

### CODE REVIEW

File: tests/js/spec/components/eventsTable/eventsTableRow.spec.jsx [hunk 1]
```diff
@@ -1,19 +1,24 @@
...
+        <tbody>
+          <EventsTableRow
+            organization={TestStubs.Organization()}
+            tagList={[]}
+            {...{orgId: 'orgId', projectId: 'projectId', groupId: 'groupId'}}
+            event={TestStubs.DetailedEvents()[0]}
+          />
+        </tbody>
+      </table>,
+      TestStubs.routerContext()
```
Comment (billyvg) [inline tests/js/spec/components/eventsTable/eventsTableRow.spec.jsx:20]: Needed, otherwise throws react warnings

File: package.json [hunk 2]
```diff
@@ -142,6 +142,7 @@
     "@storybook/addon-storysource": "^5.3.3",
     "@storybook/addons": "^5.3.3",
     "@storybook/react": "^5.3.3",
+    "@visual-snapshot/jest": "^1.0.8",
```
Comment (markstory) [inline package.json:145]: Where is the code for this? I couldn't find much about it on npm.

**Review by markstory — COMMENTED** [review]
How do image snapshots work locally? Will people get failures if image results would change?

File: package.json (hunk 2 above)
Comment (billyvg) [reply package.json:145]: Ah let me put up a repo -- deciding if all this stuff should be in a new gh org or in a mono-repo

Comment (billyvg) [reply package.json:145]: https://github.com/getsentry/jest-visual-snapshot

**Comment by billyvg** [comment]
Currently, by default, local runs do not produce any output. You can provide env vars to produce html or image snapshots. This moves snapshot comparison outside of the repo/jest and into the github action.

**Comment by markstory** [comment]
Being able to dump images locally is probably good enough for now. I'm really looking forward to not having to deal with jest snapshots going forwards :clap:

**Comment by billyvg** [comment]
@markstory did you have ideas around a local dev-ex workflow?

**Comment by markstory** [comment]
@billyvg Not good ones, as dumping image diffs in a terminal is gross, and opening a browser would be obnoxious too. I would prefer in the long term that we don't have many render 'tests'. Instead I would prefer to move the various render states out of tests and into storybook. Then add visual regressions to storybook states.

**Comment by billyvg** [comment]
You can view the rendered jest snapshot images here: https://storage.googleapis.com/sentry-visual-snapshots/getsentry/sentry/c3cb60e108b5c9d83e0dd62c936992dd116982db/index.html#added

**Comment by k-fish** [comment]
Does this launch puppeteer for each snapshot in lieu of using JSDom? It looks like it's launching from here (https://github.com/getsentry/jest-visual-snapshot/blob/main/packages/jest-environment/src/teardown.ts#L37) for each snapshot created, and isn't creating pages on an existing browser. The jests tests look to be around the same speed as master, I'm surprised there isn't more of a performance impact from launching a browser.

**Comment by billyvg** [comment]
@k-fish Yeah - there's definitely a performance impact, the numbers you're comparing are not apples to apples. Locally, it was about 20-30% slower - alternatively, we could write the html to files and have CI process html -> image, but this seemed pretty convenient for the time being.

**Comment by billyvg** [comment]
Thoughts on this @getsentry/app-frontend ?

File: tests/js/setupFramework.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+export {};
+
+declare global {
+  namespace jest {
+    interface Matchers<R> {
+      toSnapshot(): CustomMatcherResult;
+    }
+  }
+}
```
Comment (billyvg) [inline tests/js/setupFramework.ts:12]: So the `import` should grab in this type, but Cloud Build was failing (even though GHA was passing)

**Comment by k-fish** [comment]
> @k-fish Yeah - there's definitely a performance impact, the numbers you're comparing are not apples to apples. Locally, it was about 20-30% slower - alternatively, we could write the html to files and have CI process html -> image, but this seemed pretty convenient for the time being.

I really wish GH had threads for top level comments... Yeah still, even at 20% the performance impact probably isn't a big deal for this PR as Jest doesn't seem to be the slowest yet so it doesn't affect the critical path CI time. Also a concern is starting up a browser might introduce a little more instability in jest tests both locally and on CI, my (likely wrong) ballpark figure is the whole suite would open up 274 browser instances over the suite. Doing something like dumping html and having it async render was going to be my suggestion, haha. If you are willing to go to the effort of running browser instances in lambda/gcp functions then you don't have to worry about the CI waterfall time (having a separate GHA job picking up after the jest suite) either as you can be done more or less near-instantaneously.

**Comment by billyvg** [comment]
@k-fish by default, snapshots are not taken on local runs unless explicitly enabled. (This is currently not true in practice lol). Also, if my logic is right (high probability it's not), we use a singleton to launch a browser instance, so we should only have 1 browser, but multiple pages.

I think what I'd like to do is have the option to do either, I want the convenience of less configuration to generate the snapshot, but also allow the option to easily adjust how you want to allot your resources

**Comment by k-fish** [comment]
> @k-fish by default, snapshots are not taken on local runs unless explicitly enabled. (This is currently not true in practice lol). Also, if my logic is right (high probability it's not), we use a singleton to launch a browser instance, so we should only have 1 browser, but multiple pages.
> 
> I think what I'd like to do is have the option to do either, I want the convenience of less configuration to generate the snapshot, but also allow the option to easily adjust how you want to allot your resources

I missed this part, https://github.com/getsentry/jest-visual-snapshot/blob/main/packages/jest-environment/src/browser.ts#L15, you're right, it should be returning the same instance. Not sure how puppeteer works precisely, it might still run into issues of memory consumption if it doesn't close pages (you might want to close the page in the test teardown). 

Yeah I agree you should offer both so that locally you don't have to jump through so many hoops, but couldn't you setup the configuration to have it's default be convenient for development, but override that for speed in CI? Regardless I think that for this PR it's fine either way, this is mostly just asking about future plans.

**Comment by billyvg** [comment]
Oops good call, we'll want to clean up pages - I can do actual perf/benchmarks down the road and make sure my assumptions are correct.

What do you consider convenient for local dev? In my mind, snapshots should just be completely ignored in local dev since CI handles it. Unless you're debugging a snapshot in which case you can add a env var to spit out html/png.

**Comment by k-fish** [comment]
> Oops good call, we'll want to clean up pages - I can do actual perf/benchmarks down the road and make sure my assumptions are correct.

Yeah like I said, the perf of it doesn't really matter a ton imo, unless it's wildly worse perf, it's only matter of some light amount of CPU cost. I originally brought it up because I was surprised launching a browser was so fast (but I was incorrect since it's re-using the browser instance, so now I get it 😆). 

More of a concern than the perf, the instability of using a browser and having that potentially fail the suite since afaik Jest runs sequentially. I think in the long term it would perhaps be better to defer the point of failure into a parallelizable / repeatable single task (like a cloud function running the html -> image conversion) which we could add retries to. 

The reason why the perf is somewhat moot to me is that we will probably eventually want to be splitting (perhaps just simple modulo split at first) the tests into parallel suites to reduce wall time and re-run cost, at which point the per-test time is just an issue of cost, which is relatively low and likely less important than wall-time spent waiting for CI. 

> What do you consider convenient for local dev? In my mind, snapshots should just be completely ignored in local dev since CI handles it. Unless you're debugging a snapshot in which case you can add a env var to spit out html/png.

Yeah for local dev, perhaps just add another script to `package.json` that explicitly runs jest w/ visual snapshots on in case you'd like to run them locally, so you don't have to remember which env variable to add, otherwise have them off by default, but it'd be simpler locally to run just create the images directly instead of having an extra step, despite it potentially being flaky, it's less of a concern locally since you have more control. On CI you could have the extra html->image step to improve stability and performance.

**Comment by billyvg** [comment]
There are currently some snapshots that have no height, this could mean that the component is rendered funny (e.g. absolute value), or not rendering anything at all. I'm ok having these in for now and in the future removing the snapshots for blank renders (or changing tests so that it is rendering something worthwhile)

image

**Comment by billyvg** [comment]
Some updates - I've kept the existing jest snapshots in and split out the jest diff to a separate GitHub Check run (separate from acceptance). Let's evaluate this for a bit and then decide on how to proceed with jest snapshots.

I think @priscilawebdev brought up some good points re: snapshots (catching new/unused/unnecessary/missing props, un-named components), but I think those can be addressed in different ways (e.g. typescript, eslint).

**Comment by billyvg** [comment]
The irony 
image

**Review by markstory — APPROVED** [review]
Generally looks great. There are a few odd looking snapshots that we might want to clean up later:

* jest/group-actions-render-renders-correctly.png - buttons oddly sized
* jest/new-project-platform-render-should-render.png - Missing fonts.
* jest/share-issue-renders-when-busy.png - odd window size

**Comment by billyvg** [comment]
I decided to move jest snapshots back into the acceptance test runs as it would be easier to review snapshots.

I suspect we will have some flakey snapshots with the loading indicator - we should fix this by not snapshotting loading indicators, we can simply assert they exist if that is what we are trying to test (otherwise make sure we're snapshotting the correct item).

**Pull request merged**